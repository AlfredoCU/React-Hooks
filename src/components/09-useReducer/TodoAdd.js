import React, { memo } from "react";
import PropTypes from "prop-types";
import { useForm } from "../../hooks/useForm";

const TodoAdd = memo(({ handleAdd }) => {
  const [inputs, handleInputChange, resetInput] = useForm({ description: "" });
  const { description } = inputs;

  const handleSubmit = (e) => {
    e.preventDefault();

    if (description.trim().length <= 1) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false,
    };

    handleAdd(newTodo);
    resetInput();
  };

  return (
    <section className="col-5">
      <form className="form-group" onSubmit={handleSubmit}>
        <input
          type="text"
          name="description"
          placeholder="Agregar..."
          autoComplete="off"
          className="form-control"
          value={description}
          onChange={handleInputChange}
        />
        <button type="submit" className="btn btn-primary btn-block mt-1">
          Agregar
        </button>
      </form>
    </section>
  );
});

TodoAdd.propTypes = {
  handleAdd: PropTypes.func.isRequired,
};

export default TodoAdd;
