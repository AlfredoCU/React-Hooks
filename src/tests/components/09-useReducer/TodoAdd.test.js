import React from "react";
import { shallow } from "enzyme";
import "@testing-library/jest-dom";
import TodoAdd from "../../../components/09-useReducer/TodoAdd";

describe("Prueba al componente TodoAdd", () => {
  const handleAdd = jest.fn();
  const wrapper = shallow(<TodoAdd handleAdd={handleAdd} />);

  test("Foto del componente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("No debe de llamar handleAdd", () => {
    //! Simulamos el evento change, con un argumento simulado el dom.
    const value = "";
    wrapper.find("input").simulate("change", {
      target: {
        value,
        name: "description",
      },
    });

    //! Ejecutamos un submit.
    const formSubmit = wrapper.find("form").prop("onSubmit");
    formSubmit({ preventDefault() {} });

    //! No se llama la función ya que la longitud de value <= 1
    expect(handleAdd).toHaveBeenCalledTimes(0);
  });

  //! Para que funcine se debe de quitar el memo del componente.
  test("Debe de llamar la función handleAdd", () => {
    //! Simulamos el evento change, con un argumento simulado el dom.
    // const value = "Aprender React";
    // wrapper.find("input").simulate("change", {
    //   target: {
    //     value,
    //     name: "description",
    //   },
    // });
    //! Ejecutamos un submit.
    // const formSubmit = wrapper.find("form").prop("onSubmit");
    // formSubmit({ preventDefault() {} });
    //! Se llama la función handleAdd.
    // expect(handleAdd).toHaveBeenCalled(); // Se debe de llamar.
    // expect(handleAdd).toHaveBeenCalledTimes(1); // Se llama una vez.
    // expect(handleAdd).toHaveBeenCalledWith(expect.any(Object)); // Se le envia como argumento un objeto.
    // expect(handleAdd).toHaveBeenCalledWith({
    //   id: expect.any(Number),
    //   desc: value,
    //   done: false,
    // }); // Evaluamos que si sea el objeto esperado.
    //! Reset.
    // expect(wrapper.find("input").prop("value")).toBe(""); // se ejecuta la funcion reset y debe de dejar el valor en un string vacio.
  });
});
