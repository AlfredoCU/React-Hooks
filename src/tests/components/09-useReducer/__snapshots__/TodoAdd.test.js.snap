// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba al componente TodoAdd Foto del componente 1`] = `
<section
  className="col-5"
>
  <form
    className="form-group"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Agregar..."
      type="text"
      value=""
    />
    <button
      className="btn btn-primary btn-block mt-1"
      type="submit"
    >
      Agregar
    </button>
  </form>
</section>
`;
